graph/preprocessor.go

```diff
@@ -0,0 +1,211 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+/* Alias preprocessor:
+* The set of elements here are meant to process the alias definition portion of task.yaml
+* files. This is done by unmarshaling these elements which will then be added in a hierarchical
+* manner. Note the input must still be valid Yaml.
+*
+* Existing issues: Once the preTask is parsed out and the definitions are resolved, the read in
+* yaml will be processed to include the appropriate aliases, however this will include the previously
+* parsed in definitions which is not as efficient.
+*
+ */
+
+package graph
+
+import (
+	"errors"
+	"io/ioutil"
+	"net/http"
+	"regexp"
+	"strings"
+	"time"
+
+	yaml "gopkg.in/yaml.v2"
+)
+
+var (
+	errUnknownAlias              = errors.New("unknown Alias")
+	errImproperDirectiveOverload = errors.New("$ directive can only be overwritten by a single character")
+	errImproperKeyName           = errors.New("alias key names only support alphanumeric characters and _ , - , ! characters")
+	directive                    = '$'
+)
+
+// PreTask intermediate step for processing before complete unmarshall
+type PreTask struct {
+	AliasSrc  []*string         `yaml:"alias-src"`
+	AliasMap  map[string]string `yaml:"alias"`
+	directive rune
+}
+
+// Prevents recursive definitions from occuring
+func (preTask *PreTask) resolveMapAndValidate() error {
+	//Set directive from Map
+	preTask.directive = directive
+	if _, ok := preTask.AliasMap[string(directive)]; ok {
+		if len(preTask.AliasMap[string(directive)]) != 1 {
+			return errImproperDirectiveOverload
```

Comment:
If you allow override the directive, should it avoid the chars allowed for the alias name?

Reply:
I am not sure what you mean by this. I am also uncertain if allowing the override of the directive is even a good idea. I provide a escape sequence for it anyway.